### DISCUSSION TITLE
Core SQLA 1.4 string assigned FK raises NoReferencedColumnError in mutli model testing

### DISCUSSION THREAD

**Discussion by igolkotek** [sqlalchemy/sqlalchemy#8560] [Usage Questions]
Hi, 
I ask for your help in a rather simple situation: **SQLA raises NoReferencedColumnError in case of assigning column ForeignKey as a string**.

My environment:  
- Ubuntu 22.04
- SqlAlchemy 1.4.41
- Python 3.10.4
- Postgresql 13

There are 2 related tables:  

**user/models.py**
```
metadata = MetaData(schema=settings.DB_SCHEMA)

user = Table(
    "user", metadata,
    Column("user_id", Integer, primary_key=True, autoincrement=True),
    Column("password", String),
)
```
and  

**session/models.py**

```metadata = MetaData(schema=settings.DB_SCHEMA)

session = Table(
    "session", metadata,
    Column("session_id", String, primary_key=True)
    Column("user_id", Integer, ForeignKey(f"{settings.DB_SCHEMA}.user.user_id", ondelete="CASCADE", onupdate="CASCADE"), nullable=True,  index=True), 
)

```

For pytest testing I prepare conftest.py: 

```
from user.models import metadata as user_metadata  
from sessions.models import metadata as session_metadata

engine = create_engine(DB_URL, echo=echo, echo_pool=False)

with engine.connect as conn:
    with conn.begin():
        session_metadata.create_all(conn)
        user_metadata.create_all(conn)
```

The error sump is as following:

```
tests/conftest.py:55: in db_init
    session_metadata.create_all(conn)
.venv/lib/python3.10/site-packages/sqlalchemy/sql/schema.py:4917: in create_all
    bind._run_ddl_visitor(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2211: in _run_ddl_visitor
    visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
.venv/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py:524: in traverse_single
    return meth(obj, **kw)
.venv/lib/python3.10/site-packages/sqlalchemy/sql/ddl.py:827: in visit_metadata
    collection = sort_tables_and_constraints(
.venv/lib/python3.10/site-packages/sqlalchemy/sql/ddl.py:1291: in sort_tables_and_constraints
    dependent_on = fkc.referred_table
.venv/lib/python3.10/site-packages/sqlalchemy/sql/schema.py:3833: in referred_table
    return self.elements[0].column.table
.venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:1113: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
.venv/lib/python3.10/site-packages/sqlalchemy/sql/schema.py:2519: in column
    return self._resolve_column()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ForeignKey('public.user.user_id'), raiseerr = True

    def _resolve_column(self, raiseerr=True):
    
        if isinstance(self._colspec, util.string_types):
    
            parenttable, tablekey, colname = self._resolve_col_tokens()
    
            if self._unresolvable or tablekey not in parenttable.metadata:
                if not raiseerr:
                    return None
>               raise exc.NoReferencedTableError(
                    "Foreign key associated with column '%s' could not find "
                    "table '%s' with which to generate a "
                    "foreign key to target column '%s'"
                    % (self.parent, tablekey, colname),
                    tablekey,
E                   sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'session.user_id' could not find table 'public.user' with which to generate a foreign key to target column 'user_id'

.venv/lib/python3.10/site-packages/sqlalchemy/sql/schema.py:2530: NoReferencedTableError

```

Using _user.c.user_id_ instead of _'user.user_id'_ solves the problem, but I'd prefer to find the reason of this failure.

Thank you

**Comment by CaselIT** [comment]
Hi,

I don't think you should add the schema in the column name, since both tables are in the same schema. try using only `ForeignKey("user.user_id", ...)`

**Reply by igolkotek** [reply]
Thank you. 
My initial code was without schema. It raised the same exception. 

I've looked for an answer and the best I could find was adding schema. 
But with the same result.

**Reply by CaselIT** [reply]
does the above suggestion fixes the issue? Sorry but it's not clear from the above comment

**Reply by igolkotek** [reply]
No, it doesn't help

**Reply by CaselIT** [reply]
I cannot reproduce. The code below will work without issues:
```py
from sqlalchemy import MetaData, Table, Column, Integer, String, ForeignKey, create_engine

schema = "a_schema"

metadata = MetaData(schema=schema)

user = Table(
    "user",
    metadata,
    Column("user_id", Integer, primary_key=True, autoincrement=True),
    Column("password", String),
)

session = Table(
    "session",
    metadata,
    Column("session_id", String, primary_key=True),
    Column(
        "user_id",
        Integer,
        ForeignKey("user.user_id", ondelete="CASCADE", onupdate="CASCADE"),
        nullable=True,
        index=True,
    ),
)


engine = create_engine('postgresql://scott:tiger@localhost/test', echo=True, echo_pool=False)

with engine.connect() as conn:
    conn.exec_driver_sql('CREATE SCHEMA a_schema')
    metadata.create_all(conn)
```

**Reply by igolkotek** [reply]
no no.
If all the tables are in 1 file, there are no problems. My case is the tables are in different files with separate metadata, then I have to create both tables before testing

**Reply by CaselIT** [reply]
that should have been mentioned then...
I don't think it's supported using a string to link foreign keys of tables in different metadata (pinging @zzzeek to confirm).
You will either need to use the column directly instead of the string or use the same metadata for both tables

**Reply by igolkotek** [reply]
Yes, looks like that. Every metadata instance keeps data just about it's tables.
But alembic takes a list of metadata, so, maybe it could be solved one day )

**Reply by CaselIT** [reply]
That's hard to do, unless we could somehow tell to a matadata that it should consider also these other metadata, something like `Metadata(other_meta=[meta1, meta2])`.

Still may be something worth considering. What do you think mike @zzzeek ?

**Reply by zzzeek** [reply]
if you have two metadatas you link the foreign keys using the Column object as @CaselIT  said.

re: "MetaData(other_meta=...)", is that...virtually pointing to the two metadatas, or copying?  it's not clear.   it would be confusing.   just use the same MetaData to start with if you want everything in just one metadata.

**Reply by CaselIT** [reply]
virtually pointing. basically if table is not found in the meta, search it in other metas in order if they are provided. Not sure when this extended search should be done (if only then defining fk or also in other cases)

still probably not worth the effort

**Reply by zzzeek** [reply]
defeinitely something I would have done years ago but today it feels like it would be one of those things that people never use / will be confused by, IMO

**Reply by igolkotek** [reply]
Some time ago Django has moved to a declarative user_id definition in settings as a string.
I decided I could do the same trick with SLQA - this looks to be comfortable for using some models in separate projects.
Unfortunately not.

But. I could move MetaData() assigning to a project level settings file. Thus every module will have the same metadata_instance and the problem be solved.

Thank you for your time and explanations
